core/src/com/unciv/logic/map/mapunit/MapUnit.kt

```diff
@@ -1117,6 +1117,49 @@ class MapUnit : IsPartOfGameInfoSerialization {
             UniqueTriggerActivation.triggerUnique(unique, this)
     }
 
+    /**
+     * Sets the instance name should the unit be needing a historical figure name.
+     */
+    fun setHistoricalFigureName() {
+        if (hasUnique(UniqueType.CanBeAHistoricalFigure, cache.state)) {
+            // Get both the name and associated unique for the historical figure
+            val pair = getMatchingUniques(UniqueType.CanBeAHistoricalFigure, cache.state)
+                .map { it.params[0] }.distinct()
+                .flatMap {
+                    // Get all historical figure groups that match the name or a tag
+                    civ.gameInfo.ruleset.historicalFigures.values.filter {
+                        hf -> hf.name == it || hf.hasTagUnique(it, cache.state)
+                    }
+                }
+                .distinct()
```

Comment:
This second one could probably include the job of the first and total cost could even mostly balance out? ...??? 🤔

Reply:
Added the second distinct() in there because there may be multiple tags that you apply that pick up the same unique twice.